src/qs8-gemm/scalar-qd8-f16-qb4w.c.in

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$assert DATATYPE in ["QB4_F16"]
+#include <assert.h>
+$if VARIANT == "LRINTF":
+  #include <math.h>
+
+#include <xnnpack/gemm.h>
+#include <xnnpack/math.h>
+#include <xnnpack/unaligned.h>
+#include <fp16/fp16.h>
+
+
+$DATATYPE_SPEC = {"QB4_F16": "qd8_f16_qb4w"}[DATATYPE]
+$PARAMS_STRUCT = REQUANTIZATION.lower() + "_scalar" + ("_" + VARIANT.lower() if VARIANT else "")
+$PARAMS_UNION = {"QB4_F16": "xnn_f16_qc4w_minmax_params"}[DATATYPE]
+$SCALING_PARAMS = "xnn_qd8_quantization_params" if DATATYPE in ["QB4_F16"] else ""
+$XINT8_T = "int8_t"
+$MIN_F32 = "math_min_f32"
+$MAX_F32 = "math_max_f32"
+void xnn_${DATATYPE_SPEC}_gemm_minmax_ukernel_${MR}x${NR}__scalar(
+    size_t mr,
+    size_t nc,
+    size_t kc,
+    size_t bl,
+    const ${XINT8_T}* __restrict__ a,
+    size_t a_stride,
+    const void* __restrict__ w,
+    void* __restrict__ c,
+    size_t cm_stride,
+    size_t cn_stride,
+    const union ${PARAMS_UNION} params[XNN_MIN_ELEMENTS(1)],
+    const struct ${SCALING_PARAMS} quantization_params[XNN_MIN_ELEMENTS(${MR})])
+{
+  assert(mr != 0);
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
+
+  $for M in range(MR):
+    const int8_t* a${M} = (const int8_t*) ((uintptr_t) a + a_stride * ${M});
+    uint16_t* c${M} = (uint16_t*) ((uintptr_t) c + cm_stride * ${M});
+
+  kc = round_up_po2(kc, 2);
+  do {
+    $for N in range(NR):
+        const float vksum${N} = unaligned_indexed_load_f32(w, ${N});
+    w = (const float*) w + ${NR};
+
+    $for M in range(MR):
+        const int32_t vinput_zero_point${M} = quantization_params[${M}].zero_point;
```

Comment:
Nit:

```suggestion
        const float vinput_zero_point${M} = (float) quantization_params[${M}].zero_point;
```

This is somewhat similar to what I have in my neondot and neoni8mm metakernels. Also we can do something around just converting the zero_point dtype to float but someone needs to cast it so this is cast isn't too bad.